File: frontend/src/lib/chatModes.ts
/**
 * Bốn chủ đề tư vấn của trợ lý ảo. Khách bấm chọn một chủ đề trước khi gõ câu hỏi;
 * backend ghép system prompt riêng theo `id` (backend/app/chat_modes.py — hai danh sách
 * phải khớp nhau). File này không import `server-only` để cả khung chat (client) lẫn
 * trang quản trị (server) đều dùng được.
 */

export const CHAT_MODE_IDS = ['product', 'advice', 'storage', 'recipe'] as const

export type ChatMode = (typeof CHAT_MODE_IDS)[number]

export type ChatModeInfo = {
  id: ChatMode
  /** Tên trên nút chọn chủ đề. */
  label: string
  /** Tên ngắn cho tiêu đề khung chat và nhãn ở trang quản trị. */
  shortLabel: string
  /** Dòng phụ dưới tên nút, nói rõ chủ đề này trả lời được gì. */
  description: string
  /** Placeholder của ô nhập sau khi chọn. */
  placeholder: string
  /** Bong bóng trợ lý hiện ngay sau khi chọn, trước khi khách gõ. */
  intro: string
  /** Ba câu hỏi mẫu, bấm là gửi luôn. */
  examples: readonly [string, string, string]
}

export const CHAT_MODES: readonly ChatModeInfo[] = [
  {
    id: 'product',
    label: 'Giải đáp về sản phẩm',
    shortLabel: 'Sản phẩm',
    description: 'Giá, khuyến mãi, còn hàng, xuất xứ của các mặt hàng đang bán',
    placeholder: 'Hỏi về giá, khuyến mãi, còn hàng...',
    intro: 'Bạn muốn biết gì về sản phẩm của Halona? Mình tra cứu giá và tình trạng hàng ngay.',
    examples: [
      'Loại quả nào đang giảm giá?',
      'Táo nhập khẩu giá bao nhiêu?',
      'Cửa hàng có bán nước ép không?',
    ],
  },
  {
    id: 'advice',
    label: 'Tư vấn chọn hoa quả',
    shortLabel: 'Tư vấn chọn quả',
    description: 'Chọn loại quả hợp nhu cầu: biếu tặng, cho bé, ăn kiêng, theo ngân sách',
    placeholder: 'Mua cho ai, dịp gì, ngân sách bao nhiêu?',
    intro: 'Bạn cho mình biết mua cho ai, dịp gì và ngân sách khoảng bao nhiêu nhé.',
    examples: [
      'Mình cần giỏ quả biếu tặng khoảng 500.000₫',
      'Quả nào hợp cho bé 2 tuổi?',
      'Quả nào ít ngọt cho người ăn kiêng?',
    ],
  },
  {
    id: 'storage',
    label: 'Bảo quản và sử dụng',
    shortLabel: 'Bảo quản',
    description: 'Cách cất giữ, rửa, gọt và nhận biết quả chín hay hỏng',
    placeholder: 'Bạn muốn bảo quản loại quả nào?',
    intro: 'Bạn muốn bảo quản hay dùng loại quả nào? Mình hướng dẫn cách giữ tươi lâu.',
    examples: [
      'Táo để tủ lạnh được bao lâu?',
      'Vải mua về nên bảo quản thế nào?',
      'Cà chua chín để đâu cho lâu hỏng?',
    ],
  },
  {
    id: 'recipe',
    label: 'Gợi ý công thức',
    shortLabel: 'Công thức',
    description: 'Món ăn, nước ép, sinh tố từ hoa quả bạn đã mua',
    placeholder: 'Bạn có loại quả nào, muốn làm món gì?',
    intro: 'Bạn đang có loại quả nào, muốn làm nước ép, sinh tố hay món gì? Mình gợi ý công thức nhé.',
    examples: [
      'Gợi ý sinh tố từ trái cây mình vừa mua',
      'Làm nước ép táo cà chua thế nào?',
      'Món tráng miệng từ vải?',
    ],
  },
]

/** Tra thông tin chủ đề theo id; chuỗi lạ (tin cũ, dữ liệu hỏng) trả về null. */
export function chatModeInfo(id: string | null | undefined): ChatModeInfo | null {
  return CHAT_MODES.find((m) => m.id === id) ?? null
}

File: backend/app/chat_modes.py
"""Bốn chủ đề tư vấn của trợ lý ảo, mỗi chủ đề một đoạn prompt riêng.

Khách bấm chọn chủ đề trước khi gõ câu hỏi, nên Gemini không phải tự đoán khách đang
cần gì. Mỗi chủ đề có thêm `temperature` riêng: tra cứu giá cần chính xác (thấp), gợi ý
công thức cần đa dạng (cao).

Khoá của CHAT_MODES phải khớp với `ChatMode` (Pydantic kiểm ở biên API) và với
`CHAT_MODE_IDS` ở frontend/src/lib/chatModes.ts.
"""

from dataclasses import dataclass
from typing import Literal

ChatMode = Literal["product", "advice", "storage", "recipe"]


@dataclass(frozen=True)
class ChatModeSpec:
    # Nhãn tiếng Việt, dùng trong prompt và làm tag [Chủ đề: ...] khi phát lại lịch sử.
    label: str
    # Đoạn hướng dẫn riêng, nhồi vào chỗ {mode_section} của system prompt.
    section: str
    temperature: float


PRODUCT_SECTION = """\
Khách muốn hỏi về sản phẩm cửa hàng đang bán: giá niêm yết, giá đang giảm, còn hàng
hay tạm hết, thuộc danh mục nào, xuất xứ, quy cách, mô tả.
- Trả lời như tra cứu: nêu đúng tên sản phẩm, giá (và giá giảm nếu có), tình trạng
  hàng, rồi mới thêm 1 câu mô tả. Mọi con số chép nguyên từ DANH SÁCH.
- Câu hỏi chung ("có gì đang giảm giá?", "có bán xoài không?") thì liệt kê tối đa 3
  sản phẩm phù hợp nhất; không có thì nói thẳng và gợi ý mặt hàng gần nhất.
- Không suy đoán về số lượng tồn kho, ngày nhập, lô hàng hay chất lượng từng quả.
- Độ dài: 60-80 từ."""

ADVICE_SECTION = """\
Khách cần bạn chọn giúp loại quả hợp nhu cầu: biếu tặng, giỏ quà công ty, cho trẻ
nhỏ, người lớn tuổi, bà bầu, ăn kiêng, tráng miệng, ép nước, theo ngân sách.
- Nếu chưa rõ mua cho ai / dịp gì / ngân sách, hỏi lại ĐÚNG 1 câu ngắn rồi mới gợi ý.
- Gợi ý 1-3 sản phẩm có trong DANH SÁCH và CÒN HÀNG, mỗi sản phẩm kèm 1 lý do ngắn
  gắn với nhu cầu (vị, độ ngọt, dễ ăn, sang để biếu, hợp túi tiền...). Ưu tiên hàng
  đang giảm giá khi khách quan tâm ngân sách.
- Khách nêu ngân sách thì cộng nhẩm theo giá trong DANH SÁCH và nói rõ tổng ước tính.
- Kết bằng 1 câu hỏi để chốt hoặc mở rộng lựa chọn.
- Độ dài: 80-120 từ."""

STORAGE_SECTION = """\
Khách muốn biết cách bảo quản và dùng hoa quả đúng cách: để tủ lạnh hay nhiệt độ
phòng, giữ được bao lâu, cách rửa, gọt, cắt, dấu hiệu chín hoặc hỏng, mẹo giữ tươi.
- Loại quả có trong DANH SÁCH và có ghi chú "bảo quản" thì dùng đúng ghi chú đó làm ý
  đầu tiên, rồi mới bổ sung kiến thức phổ thông.
- Trình bày bằng gạch đầu dòng theo thứ tự: nơi để / nhiệt độ → thời gian dùng tốt
  nhất → mẹo hoặc lưu ý khi dùng. Nêu số ngày cụ thể khi có.
- Quả không có trong DANH SÁCH vẫn hướng dẫn được (kiến thức phổ thông), nhưng không
  nói cửa hàng đang bán quả đó. Không đưa lời khuyên y tế.
- Đây là hỗ trợ sau mua: không chào bán sản phẩm trừ khi khách hỏi.
- Độ dài: 80-120 từ."""

RECIPE_SECTION = """\
Khách muốn có công thức món ăn, nước ép, sinh tố, salad, chè, mứt... từ hoa quả.
- Ưu tiên các loại quả trong mục HOA QUẢ KHÁCH ĐÃ MUA HOẶC ĐANG CÓ TRONG GIỎ (nếu có).
  Không có mục đó và khách chưa nói mình có quả gì thì hỏi lại ĐÚNG 1 câu.
- Mỗi lần gợi ý 1-2 món. Mỗi món: tên món → nguyên liệu (định lượng cho 1-2 người) →
  3-5 bước làm ngắn gọn → 1 mẹo nhỏ (độ ngọt, đá, thời điểm dùng).
- Nguyên liệu phụ thông dụng (đường, sữa chua, mật ong, đá, muối...) được phép nêu dù
  cửa hàng không bán. Cần thêm loại quả nào mà cửa hàng có trong DANH SÁCH và còn
  hàng thì nhắc đúng tên để khách mua kèm, tối đa 2 sản phẩm.
- Độ dài: tối đa 150 từ."""

CHAT_MODES: dict[str, ChatModeSpec] = {
    "product": ChatModeSpec("Giải đáp về sản phẩm", PRODUCT_SECTION, 0.2),
    "advice": ChatModeSpec("Tư vấn chọn hoa quả", ADVICE_SECTION, 0.4),
    "storage": ChatModeSpec("Bảo quản và sử dụng", STORAGE_SECTION, 0.3),
    "recipe": ChatModeSpec("Gợi ý công thức", RECIPE_SECTION, 0.7),
}
